### ISSUE TITLE
Proposal: More control for style attribute.

### ISSUE THREAD

**Issue by sherubthakur** [mozilla/bleach#341]
Right now we can only pass in a list of css styles based on which we do the filtering. What opinion do you folks have on taking in a dict of styles instead and use it to filter out styles based on if they match or not.

i.e. right now we test cases like this would pass. (input, styles, expected)
```(
        'font-family: "Arial";',
        ['font-family'],
        'font-family: "Arial";'
    ),```

but if we can achieve something like 
```(
        'border: 1px solid blue; color: black;',
        {'font-family': '1px', 'color': 'black'},
        'color: black;'
    ),```

i.e. treating the values of the styles dict as regexs and only allowing styles that match the given regex. Support for functions would be also helpful.

Thoughts? If we do decide to do this I would be interested to work on it.

**Comment by willkg** [comment]
I'm not sure I understand where you're going with this. From what you wrote, it sounds to me like you want to enforce style attribute property *and* values. Does that sound right?

What's the use case that you're trying to solve? Is it something like "I have difficulty expressing the list of style properties and values that we want to keep"? Or "we need to clean based on style property *and* value"? Or something else?

**Comment by sherubthakur** [comment]
The use case I have in mind is I want to disallow/allow certain style values.

Why?
1. Certain styles can have values that can be used to do XSS but I do not want to outright filter them but have a regex or a function against those values which can check if these can values can lead to XSS and only filter those ones, and allow the rest.  
So from OWASP's XSS Filter Evasion Cheat Sheet background-image can be used to do XSS, but I would like to have control so that I can check `background-image` against a testing regex or a function and allow most values but block the ones that look invalid.
e.g. `<DIV STYLE="background-image: url(javascript:alert('XSS'))">`.

2. General filtering like I want to allow only certain values for some styles.

I my head it's like giving one more level of control to the user of this lib. @willkg thoughts?

**Comment by willkg** [comment]
I can't tell what the compelling need is here. Is this a "nice-to-have" sort of thing or is it a "I need this functionality in the project I'm working on"? So far everything you've written feels like "nice-to-have".

Anyhow, thinking about the API here a bit, maybe it makes sense to follow `linkify` callbacks? Can we rework the existing CSS value sanitization as a set of functions and then let developers set up the ones they want and add their own? That feels easier to document and explain, and seems easier to deal with edge cases, than supporting a dict with values.

What do you think?

**Comment by sherubthakur** [comment]
Yup I am thinking about this feature in terms of "nice-to-have". Thinking about the API I have also changed my opinion and maybe how the handling is for attributes might be useful. Here is an example just to make sure we are on the same page.

Currently this works
```
Input: 'font-family: "Arial";',
styles: ['font-family'],
Output: 'font-family: "Arial";'
```

In addition now this will also work
```
def my_styles(name, value):
    if name == 'font-family':
        return value == '1px'
    if name == 'color':
        return value == 'black'

Input: 'border: 1px solid blue; color: black;',
styles: my_styles,
Output: 'color: black;'
```

Does this make more sense and in your opinion would it be worthwhile to add it?

**Comment by willkg** [comment]
I think for the purposes of consistency, I'd make it work just like the linkify callback works.

With linkify, the callback function takes a dict of attributes and returns the adjusted list of attributes or `None` which drops the link altogether.

In this case, it seems prudent to have the callback function take a dict of style properties and return the adjusted list of style properties or `None` to drop the style attribute altogether. I think it helps to have context for the style, so maybe it should also get the tag `(namespace, tagname)` tuple as well.

So then we'd have something like this:

```
>>> def no_color_in_p(tag, properties):
...     if tag == (None, 'p') and 'color' in properties:
...         del properties['color']
...     return properties
>>> bleach.clean('<p style="color: black;">some text</p>', tags=['p'], styles=[no_color_in_p])
'<p>some text</p>'
```

Or something like that--I'm writing that by hand so there might be errors.

What do you think about that?

**Comment by sherubthakur** [comment]
I was thinking more like _User passes in a check (i.e. a pure function that only checks the value and returns True/False) and the lib deals with deleting the style attribute_. Hence the initial regex based approach.
So thinking along those lines but a better API how about a either a list (like presently implemented) or a dict (`Style -> (String -> Bool))`) Approach 1
Example
```
def color_black_only(value):
    return value == 'black'

styles={
    'font-family': lambda value: re.match('1px', value),
    'color': color_black_only,
}
```
there is a drawback however where there is no tag present (unsure how useful that would be) but if we were to support that then the type of dict changes from
dict (`Style -> ((String, Maybe Tag) -> Bool))` Approach 2
Example
```
def color_black_only(value, tag=None):
    return tag == 'p' and value == 'black'

styles={
    'color': color_black_only_on_p,
}
```
The final call would look like
bleach.clean('<p style="color: black;">some text</p>', tags=['p'], styles=styles)
'<p>some text</p>'

I find this API easier, though I am fairly inexperienced in library designing so I would delegate the API design to you haha...
So what's your pick Approach 1, Approach 2 or Linkify Approach?

**Comment by willkg** [comment]
I'd go with the linkify approach for the reasons I already stated.

When you say "better API", how are you determining what's better? Is it better support of likely use cases? If so, what are the likely use cases here?

**Comment by willkg** [comment]
This is doable with Bleach 5.0 which redid how CSS sanitizing works. You can write your own `CSSSanitizer` class that does what you need it to do.

https://bleach.readthedocs.io/en/latest/clean.html#sanitizing-css
